woocommerce/woocommerce
# [Experimental] Block structure for sharing and syncing filter blocks between main filter and overlay

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Closes #50866, replaces #51350 

This PR implements the design direction ([comment](https://github.com/woocommerce/woocommerce/issues/50866#issuecomment-2343163774)):

- Add a new template part called `Product Filter Blocks`, which is shared between `Product Filters` and `Product Filters Overlay` template parts. This template part holds the title and all filter blocks.
- In the editor, make the template part react to the overlay setting:
    - When the overlay is set to Never, the `Overlay Navigation` block is hidden.
    - When the overlay is set to Always, the `Product Filter Blocks` template part is hidden.

- On the front end, only the necessary content is rendered. For example, when the overlay setting is set to Mobile, if a shopper visits the store from a mobile device, only the trigger button and the overlay modal are rendered.

Detailed code changes:

- Refactored logic for managing inner blocks in the `product-filters/edit.tsx` using `useSelect` for better state management. Removed the `useEffect` callback and modified inner blocks only when the overlay setting was updated.
- Updated `getInnerBlockByName` function to allow a callback for flexible block matching.
- Refactored the rendering logic of the `Product Filters` block:
    - Replaced the custom template part renderer with `do_blocks`.
    - Updated to manually rendering inner blocks to control the mobile content .
    - Removed (now) unnecessary context.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.

2. Ensure the `Product Filters` and `Product Filters Overlay` template parts are reset.
3. Edit the `Product Catalog` template, add `Product Filters` template part to it.
4. Save and see the filter blocks work on the frontend.
5. From the list view, select the `Product Filters In-line` block.
6. Change the Overlay setting in the sidebar to Mobile. See the Navigation button appear.
7. Change the Overlay setting to Always, and see the `Product Filter Blocks` template part disappear.
8. Change the Overlay setting to Never, see the navigation button disappear, and the `Product Filter Blocks` template part appear again.
9. Change the Overlay setting to Mobile, and make some changes to the filter blocks. Save the changes and test on frontend.
    1. On mobile, see the trigger button only, click on it, see the overlay opened in full screen.
    2. On the desktop, don't see the trigger button, instead, see filter blocks.

10. In both cases:
    1. see the modification in step 8 applied in both main filter blocks and overlay modal.
    2. check the source and don't see unnecessary HTML loaded. On mobile, only the trigger button and modal are rendered.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
